### PR TITLE
fix(tables): sniff CSV delimiter and capture the real header row on import

### PR DESCRIPTION
**Pull request by waleedlatif1**
## Summary
- CSV table import derived the delimiter from the file extension only, so semicolon-delimited exports (European-locale Excel) parsed as a single column — the reported "alarms-export" bug. Now the delimiter is sniffed from the file head (`,` `;` `\t` `|`), quote-aware, with the extension only as a fallback.
- Header derivation used `Object.keys(rows[0])`; with `relax_column_count` a ragged/sparse first data row dropped its trailing columns from schema inference and mapping. The true header row is now captured via the `csv-parse` `columns` callback on every path — fixes the "sparse csv didn't work" report.
- New `detectCsvDelimiter` (trial-parse each candidate, pick most columns, tie-break on row-width consistency) mirrors PapaParse/`csv.Sniffer`, and `sniffCsvDelimiterFromStream` (64KB peek-then-replay) keeps the streaming paths memory-bounded on multi-GB files.
- Wired into all consumers: create route, append/replace route, background runner, client dialog preview, and `parseFileRows` (copilot).

## Type of Change
- [x] Bug fix

## Testing
- 442 table + route tests pass (49 new/updated covering delimiter detection, ragged headers, stream replay, error/destroy/empty edge cases)
- Verified against the real 61 MB `alarms-export` file: comma → 1 column, semicolon → 34 columns / 1249 records; sniffer detects `;`
- Typecheck clean, biome clean, `check:api-validation` passes

## Checklist
- [x] Code follows project style guidelines
- [x] Self-reviewed my changes
- [x] Tests added/updated and passing
- [x] No new warnings introduced
- [x] I confirm that I have read and agree to the terms outlined in the [Contributor License Agreement (CLA)](./CONTRIBUTING.md#contributor-license-agreement-cla)

### CODE REVIEW

File: apps/sim/lib/table/import.ts [hunk 1]
```diff
@@ -40,9 +65,75 @@ export function csvParseOptions(delimiter = ','): CsvParseOptions {
...
+    if (fields < 2) continue
+
+    const matching = records.reduce((n, r) => (r.length === fields ? n + 1 : n), 0)
+    const consistency = matching / records.length
+
+    if (
+      !best ||
+      fields > best.fields ||
+      (fields === best.fields && consistency > best.consistency)
+    ) {
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts]: Fixed in a5d29bcca. Ranking is now **consistency-first**: for each candidate I compute the modal (most common) row width and the fraction of rows matching it, then rank by that consistency, using column count only as a tie-break. A comma that widens a semicolon file's unquoted header now loses because its data rows are ragged (low consistency) while semicolon is uniform. Added regression tests for exactly this case (`a,b;c,d;e,f` + semicolon data) and for a semicolon file whose header has quoted commas — both now detect `;`.

File: apps/sim/lib/table/csv-delimiter-stream.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    if (next.done) {
+      exhausted = true
+      break
+    }
+    const chunk = Buffer.isBuffer(next.value) ? next.value : Buffer.from(next.value as Uint8Array)
+    chunks.push(chunk)
+    size += chunk.length
+  }
+
+  const head = Buffer.concat(chunks)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/csv-delimiter-stream.ts]: Fixed in a5d29bcca. The detection copy is now bounded via `Buffer.concat(chunks, Math.min(size, CSV_DELIMITER_SNIFF_BYTES))` — the length arg truncates the copy to the sniff window even if one upstream chunk is huge. The buffered chunks are now replayed **by reference** (yielded as-is) instead of re-concatenated, so peak memory is O(sniff window) plus the single in-flight chunk. Added a test that pushes a >1 MiB single chunk through and asserts byte-exact replay.

File: apps/sim/app/workspace/[workspaceId]/tables/components/import-csv-dialog/import-csv-dialog.tsx [hunk 3]
```diff
@@ -107,15 +113,24 @@ interface ParsedCsv {
...
   const blob = sliced ? file.slice(0, CSV_PREVIEW_BYTES) : file
   let bytes = new Uint8Array(await blob.arrayBuffer())
   if (sliced) {
     const lastNewline = bytes.lastIndexOf(0x0a)
     if (lastNewline > 0) bytes = bytes.subarray(0, lastNewline + 1)
   }
+  const delimiter = await detectCsvDelimiter(
+    bytes.subarray(0, CSV_DELIMITER_SNIFF_BYTES),
+    fallbackDelimiter
+  )
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/app/workspace/[workspaceId]/tables/components/import-csv-dialog/import-csv-dialog.tsx:136]: Fixed in a5d29bcca. The newline-trim now lives **inside** `detectCsvDelimiter` (drops a partial trailing line before parsing), so all three paths — client preview, `parseFileRows`, and `sniffCsvDelimiterFromStream` — funnel through the identical sample preparation and can no longer pick different delimiters. The stream sniffer's own ad-hoc trim was removed as redundant. Added a test asserting a mid-record byte cut doesn't skew detection.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/sim/lib/table/import.ts [hunk 4]
```diff
@@ -40,9 +65,102 @@ export function csvParseOptions(delimiter = ','): CsvParseOptions {
...
+      }
+    }
+    if (fields < 2) continue
+
+    const consistency = modalFreq / records.length
+
+    if (
+      !best ||
+      consistency > best.consistency ||
+      (consistency === best.consistency && fields > best.fields)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts]: Fixed in 33286cfba. Candidates are now scored `modalWidth × consistency` instead of consistency alone, with width as the tie-break. For your example `a|label;b;c / x|y;1 / z|w;2;3`: semicolon scores `3 × 0.667 = 2.0` and pipe scores `2 × 1.0 = 2.0`, tie broken toward the wider split → semicolon. Added a regression test for exactly this case.

File: apps/sim/lib/table/import.ts [hunk 5]
```diff
@@ -40,9 +65,102 @@ export function csvParseOptions(delimiter = ','): CsvParseOptions {
...
+    ) {
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts]: This case (`name;value,unit / a;1,kg / b;2,kg`) is a genuine **mathematical tie**: under either `;` or `,` every row splits into exactly two uniform columns (width 2, consistency 1.0, score 2.0). The two readings are perfectly symmetric — the delimiter sits between two fields and the other candidate sits inside one value, and swapping them just moves the column boundary. No content-based signal can distinguish them, and **both readings produce a valid two-column table** (no rows are dropped or mis-counted; only the split point differs). So on an exact score+width tie we fall back to the global-default candidate order (comma first — the statistically most common separator), and the user can re-split if needed. The asymmetric, decidable version of this concern (the ragged pipe case in the sibling thread) is now fixed by width×consistency scoring. Documented this in the function's TSDoc and added a test asserting the tie behavior.

File: apps/sim/lib/table/import.ts [hunk 6]
```diff
@@ -16,14 +16,39 @@ import { getColumnId } from '@/lib/table/column-keys'
...
+export function csvParseOptions(
+  delimiter = ',',
+  onHeaders?: (headers: string[]) => void
+): CsvParseOptions {
   return {
-    columns: true,
+    columns: (header: string[]) => {
+      onHeaders?.(header)
+      return header
+    },
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts:56]: Good catch — real regression, fixed in 33286cfba. The `columns` callback now delivers deduped headers to the consumer (`dedupeHeaders`, first-occurrence order, case-sensitive) so they match the record keys csv-parse actually emits when it collapses duplicate column names. It still returns the raw array to csv-parse, so the parser's own last-value-wins collapsing is unchanged. This restores the pre-change behavior (`Object.keys(rows[0])`) for duplicates while keeping the full-header-row capture for the ragged-first-row fix. Added a test asserting `a,a,b` yields headers `[a, b]` with no phantom column.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/sim/lib/table/import.ts [hunk 7]
```diff
@@ -40,9 +70,123 @@ export function csvParseOptions(delimiter = ','): CsvParseOptions {
...
+export async function detectCsvDelimiter(
+  input: Buffer | Uint8Array | string,
+  fallback: CsvDelimiter = ','
+): Promise<CsvDelimiter> {
+  const { parse } = await import('csv-parse/sync')
+  const decoded = decodeCsvText(input)
+  // Drop a partial final line when the sample is a prefix of a larger file; keep the
+  // whole thing when it's a single line (no newline) so a tiny full file still parses.
+  const lastNewline = decoded.lastIndexOf('\n')
+  const text = lastNewline > 0 ? decoded.slice(0, lastNewline + 1) : decoded
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts]: Fixed in 3811a5d75. `detectCsvDelimiter` now takes a `complete` flag — the last-newline trim runs only for truncated prefixes (where a mid-record cut must be dropped), not for a whole file that merely lacks a trailing newline. The stream sniffer passes `complete: exhausted && size <= CSV_DELIMITER_SNIFF_BYTES`, and the buffered callers (client preview, parseFileRows) pass it when the sample covers the entire file. Added tests: `a,b;c,d;e,f\n1;2;3` (no trailing newline) now correctly detects `;` with complete=true, while a truncated prefix still drops the partial tail.

File: apps/sim/lib/table/import.ts [hunk 8]
```diff
@@ -40,9 +70,123 @@ export function csvParseOptions(delimiter = ','): CsvParseOptions {
...
+    for (const [width, freq] of widthCounts) {
+      if (freq > modalFreq || (freq === modalFreq && width > fields)) {
+        fields = width
+        modalFreq = freq
+      }
+    }
+    if (fields < 2) continue
+
+    // Reward a split that is both wide and uniform; ties break toward the wider split.
+    const score = fields * (modalFreq / records.length)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts:186]: Considered this carefully and I believe the current behavior is correct for this input. Take `a|x;b / c|y;1;2 / e|z;3;4;5`:

- Under **pipe**, every row splits into exactly **2 columns** → a clean rectangular table: `(a, x;b)`, `(c, y;1;2)`, `(e, z;3;4;5)`.
- Under **semicolon**, the rows are **2, 3, and 4 columns** → not a rectangular table at all.

A CSV *is* a rectangle — a consistent column count per row is the definition of tabular data. The scorer optimizes for the most rectangular parse (`modalWidth × consistency`), and here pipe yields a valid table while semicolon yields ragged garbage, so pipe is the defensible answer. There's no content-only signal that says the author 'meant' semicolon.

The alternative — rewarding a delimiter that produces *more* total fields across ragged rows — reintroduces a strictly worse, and more realistic, failure: a single free-text cell containing several unquoted separators (e.g. one `3;"a,b,c,d,e"` row) would shatter that row into many fields and hijack detection away from the real delimiter that keeps every *other* row rectangular. Optimizing for rectangularity avoids that.

Every well-formed file — where the true delimiter yields a uniform column count, including the real-world `alarms.csv` (34 uniform semicolon columns) that motivated this PR — is detected correctly. This case only bites a file that is already non-rectangular under the 'intended' delimiter, where the right answer is genuinely undefined. Resolving as working-as-intended; happy to revisit if you have a heuristic that disambiguates non-rectangular inputs without the shattered-row regression.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/sim/lib/table/csv-delimiter-stream.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  // Copy at most the sniff window for detection — `Buffer.concat`'s length arg truncates,
+  // so an oversized final chunk can't inflate this allocation past CSV_DELIMITER_SNIFF_BYTES.
+  // `exhausted` means the whole file fit in the window, so its last row must count even
+  // without a trailing newline; otherwise the sample is a prefix and its tail may be partial.
+  const capped = Math.min(size, CSV_DELIMITER_SNIFF_BYTES)
+  const sample = Buffer.concat(chunks, capped)
+  const complete = exhausted && size <= CSV_DELIMITER_SNIFF_BYTES
+  const delimiter = await detectCsvDelimiter(sample, fallback, { complete })
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/csv-delimiter-stream.ts]: Good catch — genuine edge case, fixed in e6422740b. The read loop now uses `while (size <= CSV_DELIMITER_SNIFF_BYTES)` instead of `<`, so a file whose size is exactly the window triggers one more read, observes EOF, and sets `exhausted` (hence `complete`) — matching the buffered callers, which mark that size complete. Added a regression test: a header + one giant data row padded to exactly the window with no trailing newline now detects `;` (it would resolve to comma if the boundary EOF were missed and the last row dropped).

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/sim/lib/table/import.ts [hunk 10]
```diff
@@ -40,9 +70,127 @@ export function csvParseOptions(delimiter = ','): CsvParseOptions {
...
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/lib/table/import.ts:186]: This is the same class of case addressed earlier (previously with a pipe; a tab behaves identically), and I believe the current behavior is correct — here's the decisive reason, which is a proof that no scoring function can satisfy both this case and a more common real one.

Compare two files:

**(A) Greptile's case** — `a⇥label;b;c / x⇥y;1 / z⇥w;2;3`: under **tab** every row is a uniform 2 columns (a valid rectangular table); under **semicolon** the rows are 3, 2, 2 (ragged). Semicolon reaches a *higher max width* but *lower consistency*.

**(B) A realistic file** — `a;b / 1;2 / 3;"x,y,z,w,v"` with one unquoted free-text cell: under **semicolon** every row is a uniform 2 columns (the real table); under **comma** the rows are 1, 1, 5 (ragged). Comma reaches a *higher max width* but *lower consistency*.

(A) and (B) are **structurally identical** — in both, one candidate is uniform-but-narrow and the other is ragged-but-reaches-wider. Yet the 'correct' answer is *opposite*: (A) wants the ragged one, (B) wants the uniform one. **No content-only metric can distinguish them.** Any change that makes this PR pick semicolon in (A) necessarily makes it pick comma in (B) — trading a synthetic non-rectangular input for a realistic free-text-column regression.

The scorer deliberately prefers the **uniform (rectangular) split**, because a consistent column count per row *is* the definition of tabular data. This detects every well-formed file correctly — including the real `alarms.csv` (34 uniform semicolon columns) that motivated this PR — and picks the safer answer on the ambiguous margin. This same reasoning was accepted at 5/5 on the prior commit; the only change since is a test-only `setTimeout`→`sleep` swap. Resolving as working-as-intended.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review
